### PR TITLE
Skip laserscans with no corresponding tf transform

### LINKED ISSUE #268: [cob_scan_unifier] LookupError on start up
Appears once on startup, probably simply requires a sleep after tf::TransformListener initialization (for filling the buffer):

```
Node: /scan_unifier
Time: 16:05:39.612337463 (2016-03-04)
Severity: Error
Published Topics: /rosout, /scan_unified

Lookup would require extrapolation into the past.  Requested time 1457103936.557201524 but the earliest data is at time 1457103937.074180613, when looking up transform from frame [base_laser_front_link] to frame [base_link]

Location:
/u/fxm/git/care-o-bot/src/cob_driver/cob_scan_unifier/src/scan_unifier_node.cpp:LaserScan scan_unifier_node::unifieLaserScans:252
```

Also, `unifieLaserScans` should be `unifyLaserScans` :wink:

### PR DESCRIPTION
**Pull request by ghost**
fixes https://github.com/ipa320/cob_driver/issues/268

### CODE REVIEW

File: cob_scan_unifier/src/scan_unifier_node.cpp
Comment (mgruhler) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: could you please make this a parameter, while you are at it?

**Review by mgruhler — CHANGES_REQUESTED** [review]
Could you also add the sleep (1s) after the initialization of the transform listener, as proposed in the issue.

Also, please investigate (in simulation) what happens if you set the frequency of this node to something higher than 10 Hz.

**Comment by ghost** [comment]
Frequency was already 100Hz. Tried also with 15Hz and it works. Should I add the param to the YAMLs of all robots?

**Comment by mgruhler** [comment]
@mig-em yes, please do that.

**Comment by fmessmer** [comment]
could we also investigate #248 and #249 while @mig-em is on it?
(At least check and report whether the issues are still an issue :wink:)

**Comment by mgruhler** [comment]
@mig-em Please do as commanded ;-)

**Review by mgruhler — APPROVED** [review]

File: cob_scan_unifier/src/scan_unifier_node.cpp
Comment (benmaidel) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: please correct indentation

**Review by benmaidel — CHANGES_REQUESTED** [review]

**Comment by benmaidel** [comment]
@mig-em @ipa-mig isn't 15Hz a bad idea? See your comments here: https://github.com/ipa320/cob_robots/pull/367

I think we should switch to https://github.com/iralabdisco/ira_laser_tools because they seem to have a better implementation. I've already used their implementation once and it worked quite well.

**Comment by mgruhler** [comment]
@ipa-bnm to avoid misunderstandings: @ipa-nhg reported in #268 some problems with a rate higher then 10Hz. This should not be made the default :-)

The comment about the parameter is only to add the new parameter to the yamls...

**Comment by ghost** [comment]
So should I adapt the synchronisation part by using the TimeSynchronizer filter and investigate https://github.com/ipa320/cob_driver/issues/249 or better use the ira_laser_tools?
If I understand it right the rate param will become obsolete by using the filter.

**Comment by fmessmer** [comment]
I'd say we should evaluate `ira_laser_tools` and - if it solves all our problems - use it!
If not, fix issues in current `scan_unifier`...

**Comment by fmessmer** [comment]
Well, if `ira_laser_tools` works for us, we might want to "take over" the package and copy/move it to an ipa320 repo...as it is not released...License is [TODO](https://github.com/iralabdisco/ira_laser_tools/blob/master/package.xml#L9) :wink:

**Review by benmaidel — APPROVED** [review]

**Comment by mgruhler** [comment]
if `ira_laser_tools` is actually working good, "taking it over" would be a good idea. But I'd say we try to get in contact with the developers and see how to proceed. Last commit was Oct. 2015 for that matter...

**Comment by mgruhler** [comment]
@ipa-bnm would this be worth investigating on cob4?

**Comment by benmaidel** [comment]
@ipa-mig yes, we can/should test this on cob4 as soon as a new hardware is available

**Comment by fmessmer** [comment]
Well, would #257 have solved #268 already?

I guess we should keep #248 and #249 out of this PR...as it is separate and needs more testing.
But, I'd like to finalize/accept/close #313, #257 and #268...
@mig-em @ipa-mig @destogl

**Comment by ghost** [comment]
I tried the ira_laser_tools and they seem to work but https://github.com/ipa320/cob_driver/pull/257 and https://github.com/ipa320/cob_driver/issues/268 are not fixed (there is still the "Lookup would require extrapolation into the past." error and they use tf1).
ira_laser_merger

In this screen you see raw3-3 scans (red, green) and the unified scan (yellow)

**Comment by fmessmer** [comment]
Is the following summary correct?
@ipa-mig @mig-em @ipa-bnm 

`cob_scan_unifier`:
 - no LookupTransform error #313 #268 
 - tf2 (tested?) #257 
 - has issues merging scans #249 (were you able to reproduce #249 with your tf-fixes?)

`ira_laser_tools`
 - has LookupTransform error
 - uses tf1
 - merges scans correctly (both sim and real sensor?)
 - release issue as we are not maintainer...

**Comment by benmaidel** [comment]
Just for completeness.
There is also the possibility to use a combination of https://github.com/ros-perception/laser_assembler and https://github.com/ros-perception/pointcloud_to_laserscan to achieve a merged scan.

**Comment by ghost** [comment]
Additionally the parameters of the ira_laser_tools should be adapted.
Maybe the LookupTransform error should be fixed in the driver of the scanners, or is there a point in sending out scans without according transforms to base_link?

File: cob_scan_unifier/src/scan_unifier_node.cpp
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp:258]: @ipa-mig @mig-em 
Well, here we silently skip a scan that cannot be transformed. Should we at least notify in some way?
e.g. `ROS_ERROR_STREAM("Not using scan during unification: " << vec_laser_struct_.at(i).current_scan_msg.header.frame_id,);`

**Review by fmessmer — CHANGES_REQUESTED** [review]

**Comment by fmessmer** [comment]
@ipa-mig FYI

As discussed with @mig-em, we will try a [message_filter](http://wiki.ros.org/message_filters)-based implementation before merging this PR

**Comment by mgruhler** [comment]
good idea.

**Comment by fmessmer** [comment]
message filter implementation is in #324 
In case #324 works, this PR is obsolete

**Comment by fmessmer** [comment]
https://github.com/ipa320/cob_driver/pull/324 is the proper way to solve #268

**Pull request closed without merging**
